lighthouse-core/gather/driver.js

```diff
@@ -275,6 +276,101 @@ class Driver {
     this._nextProtocolTimeout = timeout;
   }
 
+  /**
+   * @param {LH.CrdpEvents['Target.receivedMessageFromTarget'][0]} event
+   * @param {string[]} parentSessionIds The list of session ids of the parents from oldest to youngest.
+   */
+  _handleReceivedMessageFromTarget(event, parentSessionIds) {
+    const {sessionId, message} = event;
+    /** @type {LH.Protocol.RawMessage} */
+    const protocolMessage = JSON.parse(message);
+
+    // Message was a response to some command, not an event, so we'll ignore it.
+    if ('id' in protocolMessage) return;
+
+    // We receive messages from the outermost subtarget which wraps the messages from the inner subtargets.
+    // We are recursively processing them from outside in, so build the list of parentSessionIds accordingly.
+    const sessionIdPath = parentSessionIds.concat([sessionId]);
+
+    if (protocolMessage.method === 'Target.receivedMessageFromTarget') {
+      // Unravel any messages from subtargets by recursively processing
+      this._handleReceivedMessageFromTarget(protocolMessage.params, sessionIdPath);
+    }
+
+    if (protocolMessage.method === 'Target.attachedToTarget') {
+      // Process any attachedToTarget messages from subtargets
+      this._handleTargetAttached(protocolMessage.params, sessionIdPath);
+    }
+
+    if (protocolMessage.method.startsWith('Network')) {
```

Comment:
well our current handling isn't actually on driver events it's off the connection event, and I thought it would be weird to forcibly emit an event from within driver on the connection. but I guess you disagree? 

**hopes the answer isn't let's refactor this whole thing now** ;)

Reply:
compromise: moved this bit into a function that gets called by both?

Reply:
> well our current handling isn't actually on driver events it's off the connection event

ah, I had forgotten about this. I also remembered we did it the way it is because you can't listen to `Network.*`, so we needed to listen to `protocolevent`, which is too bad.

> moved this bit into a function that gets called by both?

sounds perfect